src/array.c

```diff
@@ -479,10 +479,6 @@ JL_DLLEXPORT jl_value_t *jl_array_to_string(jl_array_t *a)
             return o;
         }
     }
-    jl_gc_count_freed(jl_array_nbytes(a));
-    a->nrows = 0;
-    a->length = 0;
```

Comment:
Pulling my comments from slack up to github:

I don't think you can safely remove these two lines, since this is a behavior change, not (just) a performance change

This would mean that sometimes `String(v::Vector)` would mutate v and sometimes it wouldn't, depending on some internal implementation details. That's not good, because it's hard to understand as a user. This is why they always set those to 0, even in the case when it's copying. So we can't change that, since it'd be a breaking change.
(That said, I completely agree with you that i've always hated this design decision. String(v) definitely shouldn't mutate v. I wish they had added a `String!(v)` that does the performance optimization, and `String(v)` would always copy. But we're left with what we have i think, and we can't make a behavior-breaking change here.

I think it would be safe to only remove the below line, and that would still achieve the affect you are after, right?:
```
    a->maxsize = 0;
```
That seems like the most correct fix:
- We are logically stealing the data out of the vector,
- even though the bytes remain owned by the vector. maxsize is not accessible from julia, only from C, so this keeps the accounting correct while preserving the existing behavior?

Reply:
Applied in the last commit.